---
title: "idxstats_cbind"
author: "Douglas Yu"
date: "28/01/2018"
output: html_document
---

This file takes the idxstats.txt outputs of samtools and bedtools (idxstats, genomecov) and combines them with the sample metadata, does some sanity checks, and removes failed samples

This code has only been checked on macOS.

```{r setup}
knitr::opts_chunk$set(
	message = FALSE,
	warning = FALSE,
	include = FALSE
)

# script-specific libraries
suppressPackageStartupMessages({
  library(stringdist)
  library(vegan)
})

# general-use packages
suppressPackageStartupMessages({
  library(tidyverse) # loads a bunch of packages (see below)
  library(here)
  library(fs)
  library(glue)
  library(readxl)
  library(cowplot)
  library(lubridate)
  library(patchwork)
  library(broom)
  library(ggeffects)
  library(viridis)
  library(arsenal) # for summary(comparedf())
  library(sjmisc) # for rotate_df()
  library(envDocument)
  library(inspectdf)
  library(conflicted)
  library(knitr)
  library(beepr)
  library(pivottabler)
  library(furrr)
  library(scales)
  library(janitor)
})

conflict_prefer("mutate", "dplyr", quiet = TRUE)
conflict_prefer("select", "dplyr", quiet = TRUE)
conflict_prefer("summarise", "dplyr", quiet = TRUE)
conflict_prefer("filter", "dplyr", quiet = TRUE)
conflict_prefer("first", "dplyr", quiet = TRUE)
conflict_prefer("here", "here", quiet = TRUE)
conflict_prefer("separate", "tidyr", quiet = TRUE)
conflict_prefer("unite", "tidyr", quiet = TRUE)
conflict_prefer("trim", "sjmisc", quiet=TRUE)
conflict_prefer("rescale", "scales", quiet=TRUE)
conflict_prefer("rescale", "discard", quiet=TRUE)
conflict_prefer("intersect", "dplyr", quiet = TRUE)
conflict_prefer("setdiff", "dplyr", quiet = TRUE) # w/out this, R crashes
conflict_prefer("to_factor", "sjmisc", quiet = TRUE)
conflict_prefer("trim", "glue", quiet = TRUE)
   
# Provide real numbers, not scientific notation.
options(scipen = 999)
```


```{r notify function}
# function to allow macOS system notifications at end of a long command (read genomecov files)
notify <- function(msgString='Message from R', titleString='Message from R', speakIt=FALSE) {
    cmd <- paste('terminal-notifier -message ', '"', msgString, '"  -title "', titleString, '"', sep='')
    system(cmd)

    if (speakIt) {
        system(paste('say', msgString))
    }
}
```


```{r set paths and filenames}
# working directory is:  "~/Dropbox/Working_docs/Luo_Mingjie_Oregon/HJA_scripts/7_idxstats_tabulate"

# samtoolsfilter <- "F2308_f0x2" # f0x2 is for PROPER PAIR, which might be too much to ask of the short 418 bp target
samtoolsfilter <- "F2308" # F2308 filter only
samtoolsqual <- "q48"
minimaprundate <- 20200929
kelpierundate <- 20200927
primer <- "BF3BR2"

# this is the enclosing folder around the idxstats and genomecov files for a given set of mappings
# outputs_F2308_f0x2_q48_minimap2_outputs_20200917_kelpie_20200916_BF3BR2_derep_filter3_vsearch97_rmdup_spikes
# the samtoolsfilter value does not change in the folder name; it is always "F2308_f0x2"
(idxstatsgenomecovfolder <- paste0("outputs_F2308_f0x2_", samtoolsqual, "_minimap2_outputs_20200929_kelpie_20200927_BF3BR2_derep_filter3_vsearch97_rmdup_spikes"))

(idxstatsgenomecovfolder_test <- glue("outputs_F2308_f0x2_{samtoolsqual}_minimap2_outputs_{minimaprundate}_kelpie_{kelpierundate}_{primer}_derep_filter3_vsearch97_rmdup_spikes"))

(outputidxstatstabulatefolder <- glue("outputs_minimap2_{minimaprundate}_{samtoolsfilter}_{samtoolsqual}_kelpie{kelpierundate}_{primer}_vsearch97"))

# this is where i store the outputs of this script
(outputidxstatstabulatefolder <- "outputs_minimap2_20200929_F2308_q48_kelpie20200927_BF3BR2_vsearch97")
dir_create(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder))

idxstatsfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted\\.bam_idxstats\\.txt")  
    # e.g. "*F2308_f0x2_q1_sorted.bam_idxstats.txt"
    # e.g. 076361-M1-S1_BDSW190602952-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt
# choose the filename suffix to match the desired samtools filter. precede with * as a wildcard. \\ is used to escape the . before txt (because it is read with grep as "any character" in list.files() below)

genomecovfile <- paste0("*", samtoolsfilter, "_", samtoolsqual, "_sorted_genomecov_d\\.txt\\.gz")  
    # e.g. ""*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz""
# choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_).      

samplemetadatafolder <- "../08_reference_sequences_datasets/"  # this is the enclosing folder around the sample metadata spreadsheet

samplemetadatafilename <- "shotgun_samples_HJAndrews_Malaisetraps_20200108.xlsx"
```


Read in all idx_stats files, add metadata to columns, and merge into one big table in long format (vertically by time)
```{r 1. read and tabulate idx files}
# make sure to unzip the output folder:  outputs_F2308_f0x2_q48_minimap2_outputs_20200917_kelpie_20200916_BF3BR2_derep_filter3_vsearch97_rmdup_spikes.tar.gz

cat("idxstatsgenomecoverfolder is: ", idxstatsgenomecovfolder)
cat("idxstatsfile search pattern is: ", idxstatsfile)
cat(genomecovfile)

idx_files <- list.files(Sys.glob(file.path("..", "..", "Kelpie_maps", idxstatsgenomecovfolder)), pattern = idxstatsfile, full.names = TRUE)

head(idx_files); cat("    ", length(idx_files), "total files")
    # Sys.glob:  [1] "outputs_F2308_f0x2_q48_minimap2_outputs_20191219_kelpie2091217_vsearch97_filtered_geneious_filtered/076361-M1-S1_BDSW190602952-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt" and so on
    # list.files is used to look for files matching "*F2308_f0x2_q1_sorted.bam_idxstats\\.txt"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt (because it is otherewise read with grep as "any character". R needs a second \)

# column names of idx files
idx_cols <- c("mitogenome", "mt_length", "mapped_reads", "unmapped_reads")

# function to read_tsv the file and extract metadata from the filename and pathname
loadFile1 <- function(x) {
    # read in the four columns and store in df
    df <- read_tsv(x,   # previously file.path(x)
    col_names = idx_cols, na = "NA",
    col_types = cols(
        mitogenome = col_character(), # calling the target sequence 'mitogenome' is from the SPIKEPIPE paper. really a barcode
        mt_length = col_integer(),
        mapped_reads = col_integer(),
        unmapped_reads = col_integer()
        ))

    # read in filename, extract and sub in the first bit of the name, remove any filepaths (using basename) and store in df$sample.  example filename:  
    # SM-PG-M1-S2_BDSW190603195-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt

  df$site <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted.bam_idxstats.txt", "\\1", basename(x), perl=TRUE)
# #     # extract site number from filename
# #     # \\w any letter, digit, underscore
# #     # \\D any non-digit, \\d any digit
# # 
    df$trap <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted.bam_idxstats.txt", "\\2", basename(x), perl=TRUE)
# #     # extract trap number from filename
# #     # \\w any letter, digit, underscore
# # 
    df$period <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted.bam_idxstats.txt", "\\3", basename(x), perl=TRUE)
# #     # extract time period from filename
# # 
    df$samtools_filter <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted.bam_idxstats.txt", "\\4", basename(x), perl=TRUE)
# #     # extract samtools filter parameters from filename
# # 
    df$mapqual_filter <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted.bam_idxstats.txt", "\\5", basename(x), perl=TRUE)
#     # extract mapping quality threshlold from filename
# 
    df$pathname <- x
#     # store full pathname in df$pathname

    # output is df
    df
 }

# lapply the loadFile() function to all of the idx_files, save the output in idx as a list
# I sometimes include package name in the command (purrr::map) when the command is generic enough that it might collide with command in another package. this is programming tic.
idx <- purrr::map(idx_files, loadFile1) # map is equivalent to lapply 
  # or could use map_dfr() instead of map followed by do.call(rbind, idx)

# Code from:  Location 6472 in Buffalo, Vince. Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools (Kindle Location 6493). O'Reilly Media. Kindle Edition.

# combine lists into a single dataframe (using do.call(rbind, idx)), reorder columns, delete unmapped_reads column, and remove rows that have "*" in the mitogenome field (the last line of each idxstats table)
idx <- do.call(rbind, idx) %>% 
    dplyr::select(mitogenome, mt_length, mapped_reads, site, trap, period, samtools_filter, mapqual_filter, pathname) %>%
    dplyr::filter(mitogenome != "*")

# sanity checks
idx %>% distinct(site) %>% count()  # there are 98 idx filenames, but there are only 96 names in the metadata Excel file
idx %>% distinct(trap) %>% count()  # 2
idx %>% distinct(trap)  # M1, M2
idx %>% distinct(pathname) %>% count()  # number of distinct values of filenames:  242  
idx %>% distinct(samtools_filter) %>% count() # number of distinct values of samtools_filter: should be 1
idx %>% distinct(samtools_filter) # should look something like:  F2308 or F2308_f0x2, depending on samtoolsfilter
idx %>% distinct(mitogenome) # 1227 (2 spikes, 1225 species)

# sanity check:  all count values should be 1285 (1283 + 2 spikes), except HOBO-357-M1-S2, which equals 2*1285 since there are two samples named HOBO-357, since one of them is Hobo-351_M1_S2, but we don't know which one, The period (S2) and trap (M1) are known, however. So we leave HOBO-357-M1-S2 untouched, and have set UTM_E, UTM_N, and Correct_Site as NA
# orig species count was 1211 when kelpie was run on only the individual samples, whereas this run includes kelpie output from nearest-neighbor groupings (each site plus its 5 nearest neighbour sites pooled into a single fastq file)
# 1225-1211 = 14 additional species
species_list <- idx %>% 
  count(site, trap, period) %>% 
  arrange(desc(n))

# idx_mitogenome_table <- idx %>% group_by(site) %>% distinct(mitogenome) %>% count() %>% arrange(desc(n)); View(idx_mitogenome_table) # sites: all columns should have 900 mitogenomes (898 + 2 spikes)
# length(idx$mitogenome) / idx_mitogenome_table[1,2] == idx %>% distinct(site) %>% count()  # should evaluate to TRUE
```

Fix errors in idxstats metadata
258355-M2-S1_BDSW190603020-1a_F2308_q48_sorted.bam_idxstats should have been named
268355-M2-S1_BDSW190603020-1a_F2308_q48_sorted.bam_idxstats (there is no site 258355)

HOBO-352-M1-S2_BDSW190603166-1a_F2308_q48_sorted.bam_idxstats.txt should have been named HOBO-353-M1-S2_BDSW190603166-1a_F2308_q48_sorted.bam_idxstats.txt (there is no site HOBO-352)

207446-M1-S1_BDSW190603001-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt should have been named 207746-M1-S1_BDSW190603001-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt (there is no site 207446)

HOBO-064-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt should have been named HOBO-060-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt (there is no site HOBO-064)

Some of the trap metadata need to be changed to the correct trap number (following the U1/2)

Some of the period metadata in the filenames have dashes (e.g. S1-1, S2-1), and these need to be changed to S1 or S2

```{r fix errors in idxstats metadata}
fixidxgenomecov <- function(df) {
  df %>% 
    mutate(
        site = case_when(
            site == "258355" & trap == "M2" & period == "S1" ~ "268355",
            site == "HOBO-352" & trap == "M1" & period == "S2" ~ "HOBO-353",
            site == "207446" & trap == "M1" & period == "S1" ~ "207746",
            site == "HOBO-064" & trap == "M1" & period == "S1" ~ "HOBO-060",
            TRUE ~ as.character(site)
        )
      ) %>% 
    mutate(
        trap = case_when(
          site == "146764" & trap == "M1" & period == "S1" ~ "M2",
          site == "161191" & trap == "M2" & period == "S2" ~ "M1",
          site == "378882" & trap == "M1" & period == "S2" ~ "M2",
          site == "HOBO-032" & trap == "M1" & period == "S2" ~ "M2",
          site == "HOBO-036" & trap == "M1" & period == "S2" ~ "M1",
          TRUE ~ as.character(trap)
        ) 
      ) %>% 
    mutate(
        period = case_when(
          site == "146764" & trap == "M1" & period == "S1-1" ~ "S1",
          site == "252447" & trap == "M2" & period == "S1-1" ~ "S1",
          site == "378882" & trap == "M1" & period == "S2-1" ~ "S2",
          site == "HOBO-032" & trap == "M1" & period == "S2-1" ~ "S2",
          site == "HOBO-036" & trap == "M1" & period == "S2-1" ~ "S2",
          TRUE ~ as.character(period)
        )
    )
}

idx <- fixidxgenomecov(idx)

idx %>% distinct(site) %>% count()  # there should be 94 idx site names, same as the 94 correct site names in the metadata Excel file

# sanity check:  all count values should be 1227, except for HOBO-357 and HOBO-036, which have ambiguous metadata
species_list <- idx %>% 
  count(site, trap, period) %>% 
  arrange(desc(n))
```

Read in sample metadata.  In the samplemetadata file, 
1.  there are two samples named HOBO-357-M1-S2, and we don't know which pair is HOBO-351-M1-S2 and which is HOBO-357-M1-S2. We thus already set site, UTM_E, UTM_N set to NA.
2.  for the two samples, HOBO-036-M1-S2-1 and HOBO-036-M1-S2, we don't know which one is period 2 and which period 1. We thus already set period for these two samples to NA. 

```{r 2. read sample metadata and make corrections}
samplemetadata <- read_excel(file.path(samplemetadatafolder, samplemetadatafilename), sheet = "shotgun_samples_HJA_Malaise_fq", na = "NA")
 
# fix errors in samplemetadata 
# Change U1/U2 in Correct_trap column to 1/2
# Add M and S prefixes to trap, period, Correct_Trap, and Correct_Period numbers
    # note that if i run this code twice, it will keep adding M and S to the columns (e.g. MM1)
# Add leading zeroes to sites 76361 and 81090, which were lost in the Excel spreadsheet
# Fix four strings in the sample column (doesn't affect the joining, since this column is not used for matching, but should fix anyway
    # 258355-M2-S1 to 268355-M2-S1
    # HOBO-352-M1-S2 to HOBO-353-M1-S2
    # HOBO-064-M1-S1 to HOBO-060-M1-S1
    # 207446-M1-S1 to 207746-M1-S1
# Fix two strings in the site column
    # HOBO-064 to HOBO-060
    # 207446 to 207746
# Keep only the rows with "pair1.truncated" (referring to the forward read of each PE read), because the samplemetadata table has a row for each fastq file (2 per sample, but i only want 1)
samplemetadata <- samplemetadata %>% 
    mutate(
        Correct_Trap = case_when(
            str_detect(Correct_Trap, "U1") == TRUE ~ "1",
            str_detect(Correct_Trap, "U2") == TRUE ~ "2",
            TRUE ~ as.character(Correct_Trap)
        )
    ) %>% 
    mutate(
        trap = str_c("M", trap),
        period = str_c("S", period),
        Correct_Trap = str_c("M", Correct_Trap),
        Correct_Period = str_c("S", Correct_Period)
    ) %>% 
    mutate(
        site = case_when(
            site == "76361" ~ "076361",
            site == "81090" ~ "081090",
            TRUE ~ as.character(site)
        )
    ) %>% 
    mutate(
        Correct_Site = case_when(
            Correct_Site == "76361" ~ "076361",
            Correct_Site == "81090" ~ "081090",
            TRUE ~ as.character(Correct_Site)
        )
    ) %>% 
    mutate(
        sample = case_when(
            sample == "258355-M2-S1" ~ "268355-M2-S1", # error in samplemetadata file
            sample == "HOBO-352-M1-S2" ~ "HOBO-353-M1-S2", # error in samplemetadata file
            sample == "HOBO-064-M1-S1" ~ "HOBO-060-M1-S1", # error in samplemetadata file
            sample == "207446-M1-S1" ~ "207746-M1-S1", # error in samplemetadata file
            TRUE ~ as.character(sample)
        )
    ) %>% 
    mutate(
        site = case_when(
            site == "207446" ~ "207746", # error in samplemetadata file
            site == "HOBO-064" ~ "HOBO-060", # error in samplemetadata file
            TRUE ~ as.character(site)
        )
    ) %>% 
    filter(str_detect(fastq_files, "pair1.truncated"))

samplemetadata %>% distinct(Correct_Site) %>% count()  # there should be 95: 94 idx site names + 1 NA (which refers to the samples that are either HOBO-357-M1-S2 or HOBO-351-M1-S2), same as the 94 correct site names in the metadata Excel file

# sanity check:  all count values should be 1, except for HOBO-036 and an NA site that is either HOBO-357-M1-S2 or HOBO-351-M1-S2, which both have count 2
species_list <- samplemetadata %>% 
  count(Correct_Site, Correct_Trap, Correct_Period) %>% 
  arrange(desc(n))

# archived code for when i want to create a new sample name from the Correct_Site/Trap/Period columns
#     unite("Correct_sample", Correct_Site, Correct_Trap, Correct_Period, sep = "_", remove = FALSE) %>% 
```


```{r 3. left_join sample metadata to idx table}
samplemetadata <- samplemetadata %>% 
  select(-site, -period, -trap) 
 
idx_meta <- left_join(idx, samplemetadata, by = c("site" = "Correct_Site", "period" = "Correct_Period", "trap" = "Correct_Trap")) %>%
  arrange(site, period, trap, mitogenome) 

names(idx_meta) 
#  [1] "mitogenome"      "mt_length"       "mapped_reads"    "site"            "trap"   
#  [6] "period"          "samtools_filter" "mapqual_filter"  "pathname"        "index"  
# [11] "fastq_files"     "sample"          "UTM_E"           "UTM_N"           "Note"   
```


```{r sanity checks}
# sanity check:  all count values should be 1227, except for HOBO-357-M1-S2 because there were two samples with this name, and for HOBO-036 because we can't differentiate the periods
species_list <- idx_meta %>% 
  count(site, period, trap) %>% 
  arrange(desc(n))

# HOBO-036-M1-S1/2 can't be differentiated, and HOBO-357-M1-S1/2 can't be differentiated, due to problems with the metadata, leaving 4 problematic samples, currently all named HOBO-036-M1-S2 and HOBO-357-M1-S2
idx_meta_table <- idx_meta %>% 
    filter(is.na(fastq_files)) %>% 
    distinct(site, period, trap)

idx_meta %>% distinct(site) %>% count()  # number of distinct values of site:  94
idx_meta %>% distinct(trap) %>% count()  # number of distinct values of trap:  should be 2 
idx_meta %>% distinct(trap) # number of distinct values of trap:  should be 2 
idx_meta %>% distinct(period) %>% count()  # number of distinct values of period:  should be 2
idx_meta %>% distinct(period)  # number of distinct values of period:  should be 2 (S1, S2) 
idx_meta %>% distinct(mitogenome) %>% count()  # number of distinct values of mitogenome: 1227 (1225 OTUs + 2 spike-ins)
idx_meta %>% distinct(mt_length)  # number of distinct values of mt_length: 894 and 897 are the two spike-ins:  full barcode plus some flanking sequence from the plasmids
idx_meta %>% distinct(samtools_filter) # %>% count()  # 1, either F2308 or F2308_f0x2
idx_meta %>% distinct(mapqual_filter) # %>% count()  # 1, q48 

idx_summ <- idx_meta %>% 
    select(-mitogenome, -mt_length, -mapped_reads, -pathname) %>% 
    distinct(site, trap, period) 
# 240 rows, which is two less than the number of sequenced samples because we can't distinguish two samples labelled HOBO-357-M1-S2 and HOBO-036-M1-S2

# write_csv(idx_summ, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "idx_summ.csv"))
``` 


```{r 4. read in genomecov_d.txt.gz files and add to idx_meta}
# genomecovfile <- "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"  # choose the filename pattern to match the desired samtools filter. precede with * as a wildcard for the sample names (e.g. Sample_IPO3916_A10_).      
# \\ is used to escape the . before txt (because it is read with grep as "any character" in list.files()

cat("Pattern search is for: ", genomecovfile)

genomecov_files <- list.files(Sys.glob(file.path("..", "..", "Kelpie_maps", idxstatsgenomecovfolder)), pattern = genomecovfile, full.names = TRUE)

head(genomecov_files); cat("    ", length(genomecov_files), "total files")
    # Sys.glob is used to generate:  
        # [1] "outputs_F2308_f0x2_q48_minimap2_outputs_20191219_kelpie2091217_vsearch97_filtered_geneious_filtered/076361-M1-S1_BDSW190602952-1a_F2308_f0x2_q48_sorted_genomecov_d.txt.gz"
    # list.files is used to look in those folders for files matching "*F2308_f0x2_q1_sorted_genomecov_d\\.txt\\.gz"
    # full.names = TRUE is used to return the pathnames
    # \\ is used to escape the . before txt and gz (because it is read with grep as "any character" in list.files()

# column names of genomecov_files
columnnames <- c("mitogenome", "position", "coverage")

# function to read_tsv the file
loadFile2 <- function(x) {
    df <- read_tsv(gzfile(x),   
        # e.g. "HOBO-064-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted_genomecov_d.txt.gz"
    col_names = columnnames, na = "NA",
    # originally, i hand-set col_types, but very large coverage numbers are returned in sci notation by bedtools and this requires col_double(). So I let R parse column types. R is able to read sci notation.
    # If i wanted to set col_types by hand, i should use col_double():
    # col_types = cols(
    #     mitogenome = col_character(),
    #     position = col_integer(),
    #     coverage = col_double()),
    trim_ws = TRUE
    )
    
# read in filename, extract and sub in the first bit of the name, remove any filepaths (using basename) and store in df$sample. example: HOBO-064-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted_genomecov_d.txt.gz
     
        df$site <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted_genomecov_d.txt.gz", "\\1", basename(x), perl=TRUE) 
        # #     # extract site number from filename 
        # #     # \\w any letter, digit, underscore 
        # #     # \\D any non-digit, \\d any digit 
 
        df$trap <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted_genomecov_d.txt.gz", "\\2", basename(x), perl=TRUE)
    # #     # extract trap number from filename
    # #     # \\w any letter, digit, underscore

        df$period <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted_genomecov_d.txt.gz", "\\3", basename(x), perl=TRUE)
    # #     # extract time period from filename
    # # 
        df$samtools_filter <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted_genomecov_d.txt.gz", "\\4", basename(x), perl=TRUE)
    # #     # extract samtools filter parameters from filename
    # # 
        df$mapqual_filter <- sub("(^[0-9, A-Z, a-z, -]+)-(M[1-2]{1})-(S[1-2,-]+)_[0-9, A-Z, a-z, -]+_([\\w]+)_(q[0-9]{2})_sorted_genomecov_d.txt.gz", "\\5", basename(x), perl=TRUE)
    #     # extract mapping quality threshlold from filename

    df$pathname_genomecov <- x
    # store full pathname in df$pathname

    # originally, we used coefficient of variation to calculate coverage, but pct_coverage is better because at high pct_coverage, CV can *increase*, causing some high coverage mappings to be deleted wrongly.  I leave the calculations of stdev and coefvar in. 
    df <- df %>% 
      dplyr::group_by(mitogenome) %>%
      summarise(sum_coverage = sum((as.integer(coverage))),
                mean_coverage = mean(as.integer(coverage)),
                stddev = sd(as.integer(coverage)),
                coefvar = sd(as.integer(coverage))/mean(as.integer(coverage)),
                length = n(),
                pct_coverage = sum(as.integer(coverage)>0)/n(), # % of positions that have 1 or more reads mapped
                site = first(site),
                trap = first(trap),
                period = first(period),
                samtools_filter = first(samtools_filter),
                mapqual_filter = first(mapqual_filter),
                pathname_genomecov = first(pathname_genomecov)
                )

    # output is df
    df
} 
 
# THIS STEP REQUIRES ~ several mins for barcodes
# lapply loadFile2() function to all of the genomecov_files, save the output in genomecovfiles as a list 
# map is equivalent to lapply 
# tictoc::tic()
# genomecoverages_summ <- purrr::map(genomecov_files, loadFile2); notify("Your loadfile2 function has finally finished!", "Message from R", speakIt=TRUE) 
# tictoc::toc()
# # rbind the list into a dataframe 
# genomecoverages_summ <- do.call(rbind, genomecoverages_summ) 


# parallel implementation
# plan(multisession) # uses all available cores, 145 secs
plan(multisession, workers = 4) # uses 4 cores, 166 secs
tictoc::tic()
genomecoverages_summ <- future_map(genomecov_files, loadFile2, .progress=TRUE); notify("Your loadfile2 function has finished!", "Message from R", speakIt=TRUE) 
tictoc::toc() # 
# rbind the list into a dataframe 
genomecoverages_summ <- do.call(rbind, genomecoverages_summ) 

# identical(genomecoverages_summ, genomecoverages_summ_furrr) # TRUE
```

Fix errors in genomecov metadata (same fixes as for the idxstats files, so i use the function fixidxgenomecov() that i wrote above)
258355-M2-S1_BDSW190603020-1a_F2308_q48_sorted.bam_idxstats should have been named
268355-M2-S1_BDSW190603020-1a_F2308_q48_sorted.bam_idxstats (there is no site 258355)

HOBO-352-M1-S2_BDSW190603166-1a_F2308_q48_sorted.bam_idxstats.txt should have been named HOBO-353-M1-S2_BDSW190603166-1a_F2308_q48_sorted.bam_idxstats.txt (there is no site HOBO-352)

207446-M1-S1_BDSW190603001-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt should have been named 207746-M1-S1_BDSW190603001-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt (there is no site 207446)

HOBO-064-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt should have been named HOBO-060-M1-S1_BDSW190603110-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt (there is no site HOBO-064)

Some of the trap metadata need to be changed to the correct trap M number (U1/2 become M1/2)

Some of the period metadata in the filenames have dashes (e.g. S1-1, S2-1), and these need to be changed to S1 or S2
```{r fix errors in idxstats metadata}
genomecoverages_summ <- fixidxgenomecov(genomecoverages_summ)

genomecoverages_summ %>% distinct(site) %>% count()  # there should be 94 idx site names, same as the 94 correct site names in the metadata Excel file

# sanity check:  all count values should be 1227, except for HOBO-357_M1_S2 and HOBO-036_M1_S2, which have two copies each (the two S2s should be S1 and S2, but we don't know which one), and thus counts of 2570
species_list <- genomecoverages_summ %>% 
  count(site, trap, period) %>% 
  arrange(desc(n))
```


```{r save genomecoverage data files, include = FALSE, eval = FALSE}
# TIME SAVING STEP  After running once, i save the file to avoid having to rerun the map() step

# F2308 samtools filter
  # saveRDS(genomecoverages_summ, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "genomecoverages_summ_F2308_minimap2_20200929_kelpie20200927.RDS"), compress = TRUE)
  # to read in genomecov files
  # genomecoverages_summ <- readRDS(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "genomecoverages_summ_F2308_minimap2_20200929_kelpie20200927.RDS"))

# F2308_f0x2 samtools filter
  # saveRDS(genomecoverages_summ, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "genomecoverages_summ_F2308_f0x2_minimap2_20200917_kelpie20200916.RDS"), compress = TRUE)
# genomecoverages_summ <- readRDS(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "genomecoverages_summ_F2308_f0x2_minimap2_20200917_kelpie20200916.RDS"))
```


```{r join genomecov and idx_meta files}
# left_join genomecoverages_summ to idx_meta
idx_meta_genomecov <- left_join(idx_meta, genomecoverages_summ) 
# R message:  Joining, by = c("mitogenome", "site", "trap", "period", "samtools_filter", "mapqual_filter")

# Only keep joins where pathname and pathname_genomecov match. 
  # 076361-M1-S1_BDSW190602952-1a_F2308_f0x2_q48_sorted.bam_idxstats.txt
  # 076361-M1-S1_BDSW190602952-1a_F2308_f0x2_q48_sorted_genomecov_d.txt.gz
idx_meta_genomecov <- idx_meta_genomecov %>% 
  mutate(
    pathname_base = str_remove(basename(pathname), "_sorted.bam_idxstats.txt"),
    pathname_gcov_base = str_remove(basename(pathname_genomecov), "_sorted_genomecov_d.txt.gz")
        ) %>% 
  filter(pathname_base == pathname_gcov_base) %>% # removes the mismatched rows
  select(-pathname_base, -pathname_gcov_base) %>% 
  arrange(site, trap, period)

# use grepl() to find mitogenome names that are COI spike names
# this regex works for both the mitogenome and barcode datasets
# my goodness, this is base R!
idx_meta_genomecov$COI_Species <- if_else(grepl("COI-SPIKE", idx_meta_genomecov$mitogenome), "COI_Spike", "Oregon_Species") 

# names(idx_meta_genomecov)
# reorder the columns for convenience
idx_meta_genomecov <- idx_meta_genomecov %>% 
  select(COI_Species, mitogenome:mapped_reads, pct_coverage, site:mapqual_filter, everything()) 

# sanity check:  all count values should be 1225 OTUs (1227 - the 2 spikes), except for HOBO-357_M1_S2 and HOBO-036_M1_S2, which have 2566 OTUs (because two samples have been pooled in each site)
species_list <- idx_meta_genomecov %>% 
  filter(COI_Species != "COI_Spike") %>% 
  count(site, trap, period) %>% 
  arrange(desc(n))

names(idx_meta_genomecov)
#  [1] "COI_Species"        "mitogenome"         "mt_length"         
#  [4] "mapped_reads"       "pct_coverage"       "site"              
#  [7] "trap"               "period"             "samtools_filter"   
# [10] "mapqual_filter"     "pathname"           "index"             
# [13] "fastq_files"        "sample"             "UTM_E"             
# [16] "UTM_N"              "Note"               "sum_coverage"      
# [19] "mean_coverage"      "stddev"             "coefvar"           
# [22] "length"             "pathname_genomecov"
```


Finally, because of the missing metadata for HOBO-036_M1_S2 and HOBO-357_M1_S2, i need to set some of the metadata for these samples to NA
HOBO-036_M1_S2 should set period to NA (could be either S1 or S2)
HOBO-357_M1_S2 should set site to HOBO-357-351 and set UTM_E & UTM_N to NA (could be either HOBO-357 or HOBO-351)

What this means for HOBO-036_M1_S1/2 is that we can use the composition information from these two samples for site, but we don't know which period each sample is.
What this means for HOBO-357-351_M1_S2 is that we can use the composition information from these two samples for period, but we don't know where these samples come from (other than from either HOBO-357 or HOBO-351). One possibility is that we will add environmental covariates that are coarse enough that they cover both sites (they are ~650 m apart). 
```{r set metadata to NA for HOBO-357_M1_S2 and HOBO-036_M1_S2}
idx_meta_genomecov <- idx_meta_genomecov %>% 
  mutate(
    period = case_when(
      site == "HOBO-036" & trap == "M1" & period == "S2" ~ NA_character_,
      TRUE ~ as.character(period)
    )
  ) %>% 
  mutate(
    site = case_when(
      site == "HOBO-357" & trap == "M1" & period == "S2" ~ "HOBO-357-351",
      TRUE ~ as.character(site)
    )
  ) %>% 
  mutate(
    UTM_E = case_when(
      site == "HOBO-357" & trap == "M1" & period == "S2" ~ NA_real_,
      TRUE ~ as.numeric(UTM_E)
    )
  ) %>% 
  mutate(
    UTM_N = case_when(
      site == "HOBO-357" & trap == "M1" & period == "S2" ~ NA_real_,
      TRUE ~ as.numeric(UTM_N)
    )
  ) 
```

Extract taxonomy information from mitogenome column. Previously, i joined a table of taxonomy information and matched by OTU name (occurrenceID, which took the form R100_1;size=100), but in the taxonomy chapter of this pipeline, i now include the taxonomy in the otu name to make it easier to see what gets mapped. Here i separate mitogenome into separate columns for OTU name and consensusClassification.

R1-124_Insecta_Lepidoptera_Ypsolophidae_Ypsolopha_canariella_BOLD_AAC0507_BLAST-EXACT-MATCH;size=29
R1063_28__Insecta_Diptera_Rhagionidae_Symphoromyia_sp_BOLD_AAP6399_size=99
R1002-21__Insecta_Psocodea_Psocidae_Loensia_NA_BOLD_ACC5693_size=385


```{r}
idx_meta_genomecov <- idx_meta_genomecov %>% 
  rename(mitogenome_orig = mitogenome) %>% 
  separate(mitogenome_orig, c("OTUname_consensusClassification", "OTUsize"), sep = ";", remove = FALSE) %>% 
  separate(OTUname_consensusClassification, c("OTUname", "class", "order", "family", "genus", "species_epithet", "BOLD", "BOLDID", "blastmatch"), sep = "_", remove = TRUE) %>% 
  unite(col = "consensusClassification", class, order, family, genus, species_epithet, BOLD, BOLDID, remove = FALSE, sep = "_") %>% 
  mutate(
    OTUname = str_c(OTUname, "_")
  ) %>% 
  unite(col = "mitogenome", OTUname, class, order, family, genus, species_epithet, BOLD, BOLDID, OTUsize, remove = TRUE)

# deprecate, because this was based on the old OTU name syntax
# idx_meta_genomecov <- idx_meta_genomecov %>% 
#   mutate(
#     mitogenome = str_replace(mitogenome, "(^R[0-9, _]+)", "\\1;")
#     ) %>% 
#   separate(mitogenome, c("OTUname", "consensusClassification", "OTUsize"), sep = ";", remove = FALSE) %>% 
#   unite(col = "mitogenome2", OTUname, OTUsize, sep = "", remove = TRUE)
# # warning message is that the COI_Spike OTUs don't match the regex and are thus ignored

# sanity check:  all count values should be 1, except for HOBO-036_M1_NA (prev HOBO-036_M1_S2) and HOBO-357-351_M1_S2 (previously HOBO-357_M1_S2), which have counts of 2
species_list <- idx_meta_genomecov %>% 
  filter(COI_Species != "COI_Spike") %>% 
  count(site, trap, period, mitogenome) %>% 
  arrange(desc(n))

# similar sanity check but using consensusClassification only.  Count values should all be 1, except for HOBO-357-351_M1_S2 and HOBO-036_M1_NA because these two sites are listed twice in the dataset. 
# I also found one consensusClassification (Insecta_Diptera_Cecidomyiidae_NA_NA_BOLD_ACV0513) that appears in two different OTUs (R10936-2 and R10941), both with size=2. Only R10936-2 got >1 mapped read (4 mapped reads) and only in one site. I keep these two OTUs, as the other one (R10941) gets filtered out later due to its low number of mapped reads
species_list <- idx_meta_genomecov %>% 
  filter(COI_Species != "COI_Spike") %>% 
  count(site, trap, period, consensusClassification, blastmatch) %>% 
  arrange(desc(n))
```

DEPRECATE  Use previous code chunk. join taxonomy file. This file adds taxonomy information for each OTU (e.g. R2052_92;size=115098). So there's no sample metadata to be fixed
```{r eval=FALSE, include=FALSE}
# kelpie_20191217_vsearch97_filtered_geneious <- read_excel(assignedtaxonomies)
# 
# idx_meta_genomecov <- left_join(idx_meta_genomecov, kelpie_20191217_vsearch97_filtered_geneious, by = c("mitogenome" = "occurrenceId")) %>% 
#   unite("site_trap_period", site, trap, period, remove = FALSE) %>% 
#   select(COI_Species, mitogenome, consensusClassification, site_trap_period, everything())
# 
# # sanity check:  all count values should be 1, except for HOBO-357-351_M1_S2 (previously HOBO-357_M1_S2) and HOBO-036_M1_NA (prev HOBO-036_M1_S2), which have counts of 2 
# species_list <- idx_meta_genomecov %>% 
#   filter(COI_Species != "COI_Spike") %>% 
#   count(site, trap, period, consensusClassification) %>% 
#   arrange(desc(n))
```


Filter out failed samples. There should always be mapped reads for the COI_spike species, so if this number is zero, the sample has failed for our purposes and should be deleted. I can set the COI_SPIKE threshold higher, but for now, i set threshold to 1.
```{r}
# calculate sum of all COI spike reads for each sample. 
sum_idx_meta_genomecov <- idx_meta_genomecov %>%
    group_by(site, trap, period, COI_Species) %>%
    summarise_at(vars(mapped_reads), ~sum(as.numeric(.))) %>%
    arrange(COI_Species, mapped_reads, site, trap, period) %>% 
  unite(col = site_trap_period, site, trap, period, remove = FALSE)

# visual inspection reveals 1 sample that has 0 COI_spike mapped reads. 
sample_names <- sum_idx_meta_genomecov %>% 
    dplyr::filter(COI_Species == "COI_Spike" && mapped_reads < 1) %>% 
    dplyr::select(site_trap_period) %>% 
    distinct(site_trap_period); View(sample_names)

samples_to_remove <- sample_names %>% 
  dplyr::pull(site_trap_period) # pull out a single variable
samples_to_remove # 1 sample to remove: SM-05_M1_S1

idx_meta_genomecov <- idx_meta_genomecov %>% 
  unite(col = site_trap_period, site, trap, period, remove = FALSE) %>% 
    filter(!site_trap_period %in% samples_to_remove) 

# sanity check:  239 samples, which is correct since i started with 240 samples, and i removed 1 sample. The reason i started with 240 samples, and not 242, is because 2 samples (HOBO-351/7_M1_S2 and HOBO-036_M1_S1/2) were pooled because of missing metadata
idx_meta_genomecov %>% 
    distinct(site_trap_period) %>% 
    count()  # number of distinct values of site_trap_period
```


DEPRECATE. THIS WILL BE DONE IN THE FSL CORRECTION SCRIPT Spike-in correction, creating a new column mapped_reads_corr
```{r, eval=FALSE, include=FALSE}
coispike <- idx_meta_genomecov %>% 
  filter(COI_Species == "COI_Spike") %>% 
  select(mitogenome, site_trap_period, site, trap, period, mapped_reads, pct_coverage) %>% 
  group_by(site_trap_period) %>% 
  summarise(mapped_reads = sum(mapped_reads)) 

# changed code to keep the 2 spike-ins in the dataset, so that they can be used downstream. idx_meta_genomecov_corr still corrects the OTUs by the sum of the spike-ins per sample, but if we want, we retain option to do a different correction. 
idx_meta_genomecov_sp <- idx_meta_genomecov 
# idx_meta_genomecov_sp <- idx_meta_genomecov %>% filter(COI_Species == "Oregon_Species")

idx_meta_genomecov_corr <- idx_meta_genomecov_sp %>% 
  left_join(coispike, by = "site_trap_period") %>% 
  select(site_trap_period, mitogenome, consensusClassification, mt_length, mapped_reads = mapped_reads.x, spike_reads = mapped_reads.y, everything()) %>% 
  mutate(
    mapped_reads_corr = round (100*(mapped_reads / spike_reads), digits = 2)
  ) %>% 
  select(site_trap_period, mitogenome, consensusClassification, mt_length, mapped_reads, mapped_reads_corr, spike_reads, everything())
```


Sanity checks and save idx_meta_genomecov_corr output file
```{r save idx_meta_genomecov_corr}
# sanity checks
# these tables check that i only have one row for each unique value of site_trap_period, and there should be 239 rows (after having removed the sample that lacks COI_SPIKE mapped reads)

# ideally, there should be a 1 in all rows of column "n"
idxstats_by_site_trap_period <- idx_meta_genomecov %>% 
    dplyr::group_by(site_trap_period) %>% 
    dplyr::distinct(site_trap_period) %>% 
    dplyr::count() %>% 
    dplyr::arrange(desc(n)); View(idxstats_by_site_trap_period)

# idxstats_by_site_trap_period <- idx_meta_genomecov_corr %>% 
#     dplyr::group_by(site_trap_period) %>% 
#     dplyr::distinct(site_trap_period) %>% 
#     dplyr::count() %>% 
#     dplyr::arrange(desc(n)); View(idxstats_by_site_trap_period)

# ideally, there should be 1227 in all rows of column "n" 
# 1227 == 1225 species + 2 spike-in species
idxstats_by_mitogenome <- idx_meta_genomecov %>% 
    dplyr::group_by(site_trap_period) %>% 
    dplyr::distinct(mitogenome) %>% 
    dplyr::count() %>% 
    dplyr::arrange(desc(n)); View(idxstats_by_mitogenome)

# idxstats_by_mitogenome <- idx_meta_genomecov_corr %>% 
#     dplyr::group_by(site_trap_period) %>% 
#     dplyr::distinct(mitogenome) %>% 
#     dplyr::count() %>% 
#     dplyr::arrange(desc(n)); View(idxstats_by_mitogenome)

# F2308 samtools filter
# write_tsv(idx_meta_genomecov_corr, gzfile(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "idx_meta_genomecov_F2308_minimap_20200929_kelpie20200927.txt.gz"))) # output file
# F2308_f0x2 samtools filter
# write_tsv(idx_meta_genomecov_corr, gzfile(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "idx_meta_genomecov_corr_F2308_f0x2_minimap_20200916_kelpie20200917.txt.gz"))) # output file


# to import the output datafile
# F2308 samtools filter
# idx_meta_genomecov_corr <- read_delim(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "idx_meta_genomecov_F2308_minimap_20200929_kelpie20200927.txt.gz"), "\t", escape_double = FALSE,
#    col_types = cols(
#      Correct_Period = col_character(),
#      Correct_Site = col_character(),
#      Correct_Trap = col_character()),
#    trim_ws = TRUE)

# F2308 f0x2 samtools filter
# idx_meta_genomecov_corr <- read_delim(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "idx_meta_genomecov_corr_F2308_f0x2_minimap_20200916_kelpie20200917.txt.gz"), "\t", escape_double = FALSE,
#    col_types = cols(
#      Correct_Period = col_character(),
#      Correct_Site = col_character(),
#      Correct_Trap = col_character()),
#    trim_ws = TRUE)
```


Make sample X species OTU table, with mapped_reads_corr, which are spike-corrected correlations and with mapped_reads, which are not spike-corrected. These tables will be joined with the environmental covariate data.
```{r}
# filter out rows without confident mappings
idx_meta_genomecov_filtered <- idx_meta_genomecov %>% 
  filter(mapped_reads > 0 & pct_coverage > 0.5) %>% 
  arrange(site_trap_period, desc(mapped_reads))

# idx_meta_genomecov_filtered <- idx_meta_genomecov_corr %>% 
#   filter(mapped_reads_corr > 0 & pct_coverage > 0.5) %>% 
#   arrange(site_trap_period, desc(mapped_reads_corr))

# sanity check:  all count values should be 1, except for HOBO-036_M1_NA and HOBO-357-351_M1_S2, which have count values of 2
species_list <- idx_meta_genomecov_filtered %>% 
  # filter(COI_Species != "COI_Spike") %>% # not needed
  count(site, trap, period, mitogenome) %>% 
  arrange(desc(n))

total_detections_by_period <- idx_meta_genomecov_filtered %>% 
  filter(COI_Species != "COI_Spike") %>% 
  count(period) %>% 
  arrange(desc(n)); View(total_detections_by_period)

mapped_reads_by_period <- idx_meta_genomecov_filtered %>% 
  filter(COI_Species != "COI_Spike") %>% 
  group_by(period) %>% 
  summarise(sum_mapped_reads = sum(mapped_reads)) %>% 
  arrange(desc(sum_mapped_reads)); View(mapped_reads_by_period)

# if i want to see the number of mapped reads per some taxon
mapped_reads_by_period_taxon <- idx_meta_genomecov_filtered %>% 
  filter(COI_Species != "COI_Spike") %>% 
  filter(grepl("Dolichovespula", mitogenome)) %>%
  group_by(period) %>% 
  summarise(sum_mapped_reads = sum(mapped_reads)) %>% 
  arrange(desc(sum_mapped_reads)); View(mapped_reads_by_period_taxon)

names(idx_meta_genomecov_filtered)

# make sample X OTU table, using pivot_wider() 
# must remove HOBO-036_M1_NA, HOBO-357-351_M1_S2 because they identify two different samples, which causes the output to be filled with list_cols

# using mapped_reads (not spike-in corrected)
idx_meta_genomecov_filtered_wide <- idx_meta_genomecov_filtered %>% 
  filter(site_trap_period != "HOBO-036_M1_NA") %>% 
  filter(site_trap_period != "HOBO-357-351_M1_S2") %>% 
  select(mitogenome, mapped_reads, site_trap_period, UTM_E, UTM_N) %>%
  mutate(
    mitogenome = str_replace_all(mitogenome, ";", "_")
  ) %>% 
  pivot_wider(
    names_from = mitogenome,
    values_from = mapped_reads,
    values_fill = list(mapped_reads = 0)
    )  

# to confirm that COI-SPIKE OTUS ARE PRESENT
names <- names(idx_meta_genomecov_filtered_wide) %>% as_tibble()

# no longer doing this because FSL correction is done in the next stage
# using mapped_reads_corr (spike-in corrected)
# idx_meta_genomecov_corr_filtered_wide <- idx_meta_genomecov_filtered %>% 
#   filter(site_trap_period != "HOBO-036_M1_NA") %>% 
#   filter(site_trap_period != "HOBO-357-351_M1_S2") %>% 
#   select(mitogenome, mapped_reads_corr, site_trap_period, UTM_E, UTM_N) %>%
#   mutate(
#     mitogenome = str_replace_all(mitogenome, ";", "_")
#   ) %>% 
#   pivot_wider(
#     names_from = mitogenome,
#     values_from = mapped_reads_corr,
#     values_fill = list(mapped_reads_corr = 0)
#     )  

```


Load environmental covariate data
envdataGIS column descriptions are in "10_eo_data/biodiversity_site_info_GIS_vars_descriptions_20200214.xlsx
envdataLidar column descriptions are in 10_eo_data/biodiversity_site_info_lidar_descriptions_20200518_2.rtf
envdataMS column descriptions are in 10_eo_data/biodiversity_site_info_multispectral_vars_2020-04-13.txt

```{r load env data and format}
envdataGIS <- read_csv(file.path("..", "10_eo_data", 
                                "biodiversity_site_info_GIS_vars_20200220.csv"))

envdataGIS <- envdataGIS %>% 
  select(SiteName, UTM_E, UTM_N, oldGrowthIndex = o, clearcut = TREATMENT_,
         elevation_m = e, canopyHeight_m = ht, minT_annual = mint, 
         maxT_annual = maxt, precipitation_mm = p, distToRoad_m = r.d, 
         distToStream_m = s.d, YrsSinceDist, insideHJA = hja) %>% 
  mutate(
    SiteName = case_when(
      SiteName == "81090" ~ "081090",
      SiteName == "76361" ~ "076361",
      TRUE ~ as.character(SiteName)
    )
  ) %>% 
  mutate(
    insideHJA = case_when(
      insideHJA == 1 ~ "yes",
      insideHJA == 0 ~ "no",
      TRUE ~ as.character(insideHJA)
    )
  ) %>% 
  mutate(
    clearcut = case_when(
      clearcut == "Clearcut" ~ "yes",
      TRUE ~ "no"
    )
  ) %>%
  filter(SiteName != "290415" & SiteName != "HOBO-040") 
# SiteName 290415 was a place where the malaise trap was stolen twice, and SiteName Hobo-040 was attacked by bears in both sessions, so these rows do not have a match with idx_meta_genomecov species data (i.e. there were no samples to sequence)

envdataMS <- read_csv(file.path("..", "10_eo_data", "biodiversity_site_info_multispectral_20200413.csv")) %>% 
  select(SiteName, UTM_E, UTM_N, starts_with("mean.")) %>% 
      mutate(
        SiteName = case_when(
            SiteName == "76361" ~ "076361",
            SiteName == "81090" ~ "081090",
            TRUE ~ as.character(SiteName)
        )
      )

envdataLidar <- read_csv(file.path("..", "10_eo_data", 
                    "biodiversity_site_info_lidar_20200518_cleaned.csv")) %>% 
  select(-l_Cover_4m_16m_all) # some values are NA

# join MS and Lidar datasets
envdataMSLidar <- full_join(envdataMS, envdataLidar, 
                            by = c("SiteName", "UTM_E", "UTM_N")) %>% 
  filter(SiteName != "290415" & SiteName != "HOBO-040") 


# full environmental covariate data
envdataGISMSLidar <- full_join(envdataGIS, envdataMSLidar, 
                            by = c("SiteName", "UTM_E", "UTM_N")) 
```


join environmental covariate data to idx_meta_genomecov_corr_filtered_wide and to idx_meta_genomecov_filtered_wide, and write output. 
```{r sample by species tables}
# mapped_reads only
# correct UTM_E and UTM_N values to the ones in the GISMSLidar tables
idx_meta_genomecov_filtered_wide <- idx_meta_genomecov_filtered_wide %>% 
  separate(site_trap_period, into = c("site", "trap", "period"), 
           sep = "_", remove = FALSE) %>% 
  select(site, trap, period, everything()) %>% 
  mutate(
    UTM_E = case_when(
      site_trap_period == "335225_M1_S2" ~ 558228,
      TRUE ~ as.numeric(UTM_E)
      )
    ) %>% 
  mutate(
    UTM_N = case_when(
      site_trap_period == "335225_M1_S2" ~ 4897232,
      TRUE ~ as.numeric(UTM_N)
    )
  )

# join idx_meta_genomecov_filtered_wide and envdataGISMSLidar
idx_meta_genomecov_filtered_wide <- envdataGISMSLidar %>% 
  full_join(idx_meta_genomecov_filtered_wide, by = c("SiteName" = "site", "UTM_E", "UTM_N")) %>%
  relocate(SiteName, UTM_E, UTM_N, period, trap, site_trap_period) 

# to see number of reads per sample for a given taxon
species <- select(idx_meta_genomecov_filtered_wide, contains(c("Vespidae"))) %>% colSums(); tibble::view(species)

# write output
# F2308 samtools filter
# write_csv(idx_meta_genomecov_filtered_wide, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_table_F2308_minimap2_20200929_kelpie20200927.csv"))

# F2308_f0x2 samtools filter
  # write_csv(idx_meta_genomecov_filtered_wide, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_table_F2308_f0x2_minimap2_20200916_kelpie20200917.csv"))

# read_csv() commands
  # mapped_reads and F2308 samtools filter
  # idx_meta_genomecov_filtered_wide <- read_csv(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_table_F2308_minimap2_20200929_kelpie20200927.csv"))
    
  # mapped_reads and F2308_f0x2 samtools filter
  # not yet created

# prev code,deprecate
# idx_meta_genomecov_corr_filtered_wide <- envdataTosa %>% 
#   full_join(idx_meta_genomecov_corr_filtered_wide, by = c("SiteName" = "site")) %>%
#   select(SiteName, UTM_E = UTM_E.x, UTM_N = UTM_N.x, everything()) %>% 
#   select(-rownumber, -UTM_E.y, -UTM_N.y)
```

Deprecated because FSL correction is now done in the next stage
```{r sample by species tables for spike corrected code, eval=FALSE, include=FALSE}
# mapped_reads_corr only
# join idx_meta_genomecov_corr_filtered_wide and envdataTosa
# idx_meta_genomecov_corr_filtered_wide <- idx_meta_genomecov_corr_filtered_wide %>% 
#   separate(site_trap_period, into = c("site", "trap", "period"), 
#            sep = "_", remove = FALSE) %>% 
#   select(site, trap, period, everything()) %>% 
#   mutate(
#     UTM_E = case_when(
#       site_trap_period == "335225_M1_S2" ~ 558228,
#       TRUE ~ as.numeric(UTM_E)
#       )
#     ) %>% 
#   mutate(
#     UTM_N = case_when(
#       site_trap_period == "335225_M1_S2" ~ 4897232,
#       TRUE ~ as.numeric(UTM_N)
#     )
#   )
# the UTM numbers for this sample in shotgun_samples_HJAndrews_Malaisetraps_20200108.xlsx are different from the UTM numbers in the 3 environmental covariate datasets, so i use the latter values

# idx_meta_genomecov_corr_filtered_wide <- envdataGISMSLidar %>% 
#   full_join(idx_meta_genomecov_corr_filtered_wide, by = c("SiteName" = "site", "UTM_E", "UTM_N")) %>%
#   relocate(SiteName, UTM_E, UTM_N, period, trap, site_trap_period) 



# write output
# F2308 samtools filter
# write_csv(idx_meta_genomecov_corr_filtered_wide, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_corr_table_F2308_minimap2_20200916_kelpie20200917.csv"))
# F2308_f0x2 samtools filter
# write_csv(idx_meta_genomecov_corr_filtered_wide, file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_corr_table_F2308_f0x2_minimap2_20200916_kelpie20200917.csv"))

# read_csv() commands 
  # mapped_reads_corr
    # F2308 samtools filter
    # idx_meta_genomecov_corr_filtered_wide <- read_csv(file.path("..", "..", "Kelpie_maps", outputidxstatstabulatefolder, "sample_by_species_corr_table_F2308_minimap2_20200916_kelpie20200917.csv"))
    # F2308 f0x2 samtools filter
    # not yet created


```



pivot table of sample, trap, period
```{r}
sampletable <- idx_meta_genomecov_filtered_wide %>% 
  select(site_trap_period) %>% 
  separate(site_trap_period, into = c("site", "trap", "period"), sep = "_", remove = TRUE)

pt <-  PivotTable$new()
pt$addData(sampletable)
pt$addColumnDataGroups("period")
pt$addColumnDataGroups("trap")
pt$addRowDataGroups("site")
pt$defineCalculation(calculationName="Count", summariseExpression="n()")
pt$evaluatePivot()
sampletable_pt <- pt$asDataFrame(separator="_")
names(sampletable_pt)
sampletable_pt <- sampletable_pt %>% 
  # rownames_to_column(var = "site") %>% 
  # filter(site !="Total") %>% 
  arrange(desc(S1_Total)); View(sampletable_pt)

# other summarising code
# library(janitor)
# tabyl(idx_meta_genomecov_filtered_wide, SiteName, trap, period, show_na=FALSE, show_missing_levels = FALSE)
# idx_meta_genomecov_filtered_wide %>% distinct(SiteName) %>% count()
# species_list <- idx_meta_genomecov_filtered_wide %>%
#   filter(trap=="M1" & period=="S1") %>%
#   count(trap, period) %>%
#   arrange(desc(n))
# species_list <- idx_meta_genomecov_filtered_wide %>%
#   count(period, trap) #%>%
#   arrange(desc(n))
```


### lulu.  
I ran lulu on the kelpie_20200214_BF3BR2_derep_filtered_geneious_vsearch97_min2.fas OTUs, and lulu removed no OTUs.  So it looks like Kelpie + vsearch97 produces separate OTUs, and lulu does not help.

prepare dataframe for lulu: uses sjmisc::rotate_df
```{r}
names(idx_meta_genomecov_filtered_wide)
idx_meta_genomecov_filtered_wide_lulu <- idx_meta_genomecov_filtered_wide %>% 
  select(site_trap_period, contains("__")) %>% 
  filter(!is.na(site_trap_period)) %>% # remove 290415 and Hobo-040, which have no data
  mutate(
    site_trap_period = str_c("x", site_trap_period)
  ) %>% 
  rotate_df(cn = TRUE) # leave species names in row.names
```


```{bash vsearch for matchlist}
primer="BF3BR2" # "LERAY"
kelpiedate="20200927"

cd ~/Dropbox/Working_docs/Luo_Mingjie_Oregon/HJA_analyses_Kelpie/HJA_scripts/05_taxonomy/kelpie_${kelpiedate}_${primer}/


vsearch --usearch_global kelpie_${kelpiedate}_${primer}_derep_filter3_vsearch97_rmdup_spikes.fas --db kelpie_${kelpiedate}_${primer}_derep_filter3_vsearch97_rmdup_spikes.fas --self --id .84 --iddef 1 --userout match_list.tsv -userfields query+target+id --maxaccepts 0 --query_cov .9 --maxhits 10
```


```{r}
primer <- "BF3BR2" # "LERAY"
kelpiedate <- 20200927
glue("kelpie_{kelpiedate}_{primer}") # check

match_list <- read_delim(file.path("..", 
                                   "05_taxonomy",
                                   glue("kelpie_{kelpiedate}_{primer}"),
                                   "match_list.tsv"
                                   ), 
                         "\t", 
                         escape_double = FALSE, 
                         col_names = FALSE, 
                         trim_ws = TRUE
                         )

curated_result <- lulu::lulu(idx_meta_genomecov_filtered_wide_lulu,
                             match_list)
curated_table <- rownames_to_column(curated_result$curated_table, 
                                    var = "OTU")
curated_result[["discarded_otus"]]
otu_map <- rownames_to_column(curated_result$otu_map, var = "OTU") %>%
  arrange(curated)
```

lulu finds no parent-child OTUs in idx_meta_genomecov_filtered_wide


### species number
```{r}
species_idx_meta_genomecov <- rotate_df(idx_meta_genomecov_filtered_wide_lulu, rn = "SiteName")

species_idx_meta_genomecov <- species_idx_meta_genomecov %>% select(-SiteName) 

specnumber(species_idx_meta_genomecov, MARGIN = 1) # number of species per site

species <- as.data.frame(specnumber(species_idx_meta_genomecov, MARGIN = 2)) %>%
  rownames_to_column(var = "species") %>% # number of species per site
  rename(incidences = "specnumber(species_idx_meta_genomecov, MARGIN = 2)") %>%
  arrange(desc(incidences)) %>% 
  filter(!grepl("COI-SPIKE", species))
View(species)

species_min_incidences <- species %>% 
  filter(incidences > 4)

otuhist <- hist(species$incidences, breaks = 1162, xlim = c(0, 100))
text(otuhist$mids, otuhist$counts, cex = 0.5, 
     otuhist$counts, adj = c(.5, -.5), col = "blue3")
```



makeblastdb does not work because info lines are too long
```{bash blastn for matchlist, include = FALSE, eval = FALSE}
#First produce a blastdatabase with the OTUs
# makeblastdb -in kelpie_20191217_vsearch97_filtered_geneious.xlsx.fasta -parse_seqids -dbtype nucl
# Then blast the OTUs against the database
# blastn -db kelpie_20191217_vsearch97_filtered_geneious.xlsx.fasta -outfmt '6 qseqid sseqid pident' -out match_list.txt -qcov_hsp_perc 80 -perc_identity 84 -query kelpie_20191217_vsearch97_filtered_geneious.xlsx.fasta
```


